### DISCUSSION TITLE
IsGranted attribute couldn't find subject when using a ValueArgumentResolver

### DISCUSSION THREAD

**Discussion by tchapuis** [symfony/symfony#48639] [Q&A] [answered]
### Symfony version(s) affected

6.2.*

### Description

I'm upgrading my project from Symfony 6.1 to 6.2 and I'm trying to remove the SensioFrameworkExtraBundle, to use the `IsGranted` attribute available in core.

So I have this controller : 

```php
#[Route('/v1/user/{userId}')]
#[IsGranted(Voter::ACCESS, subject: 'userId', statusCode: Response::HTTP_FORBIDDEN)]
class UserController extends AbstractController
{
    #[Route(methods: ['GET'])]
    public function getOneUser(User $user): Response
    {
        // Some code
    }
}
```

The `User` in the `getOnUser` method is resolved using a custom resolver implementing the `ValueResolverInterface`.

```php
class UserResolver implements ValueResolverInterface
{
    public function __construct(
        private readonly UserRepositoryInterface $userRepository
    ) {
    }

    public function resolve(Request $request, ArgumentMetadata $argument): Iterator
    {
        if ($argument->getType() !== User::class && !$request->attributes->has('userId')) {
            return [];
        }

        $userId = $request->attributes->get('userId');

        try {
            $user = $this->userRepository->get($userId);
        } catch (UserNotFoundException $e) {
            throw new NotFoundHttpException($e->getMessage(), $e);
        }

        yield $user;
    }
}
```

But I get this error : `Could not find the subject "userId" for the #[IsGranted] attribute. Try adding a "$userId" argument to your controller method.`

I can fix this issue by adding a `$userId` argument in the method but it will be unused (and my code quality tools will try to remove them)

This piece of code was working fine before updating, using the SensioFrameworkExtraBundle.

### How to reproduce

Create a controller using the `IsGranted` attribute with a Voter. The subject of the Voter should be a parameter of the route.

Create a custom Argument Resolver and use it in the arguments of a method in the controller. The argument should not have the same name as the Voter subject (or route param)

**Comment by HypeMC** [comment, marked as answer]
The `#[IsGranted]` attribute works a little differently than the SensioExtraBundle one. Try:

```php
#[IsGranted(Voter::ACCESS, subject: new Expression('request.attributes.get("userId")'), statusCode: Response::HTTP_FORBIDDEN)]
```

The subject must be a method argument or expression, but since you don't have an argument named `userId` it fails. Another option would be to add the argument:

```php
public function getOneUser(User $user, int $userId): Response {}
```

**Reply by abdellahrk** [reply]
Thank you @HypeMC. 
This has been life-saving for me.

**Comment by tchapuis** [comment]
Thank you for you answer. Both solution are working but I tend to prefer the first one. I think adding an unused argument is not a good practice, in my case.
